include/adc.h

```diff
@@ -27,119 +27,152 @@ extern "C" {
- * specified device.
+ * @brief Set a read request.
  *
- * @param dev Pointer to the device structure for the driver instance.
+ * @param config Pointer to a valid adc_config structure instance.
+ * @param sequence Pointer to an array with definitions of samplings to perform.
+ * @param sequence_length Number of entries in the sampling sequence array.
  *
- * @return N/A
+ * @returns 0 on success, negative errno code otherwise.
  */
-__syscall void adc_enable(struct device *dev);
+__syscall int adc_read(const struct adc_config *config,
+		       const struct adc_sampling *sequence,
+		       u8_t sequence_length);
 
-static inline void _impl_adc_enable(struct device *dev)
+static inline int _impl_adc_read(const struct adc_config *config,
+				 const struct adc_sampling *sequence,
+				 u8_t sequence_length)
 {
-	const struct adc_driver_api *api = dev->driver_api;
+	const struct adc_driver_api *api = config->dev->driver_api;
 
-	api->enable(dev);
+	return api->read(config, sequence, sequence_length);
 }
 
+#ifdef CONFIG_ADC_ASYNC
 /**
- * @brief Disable ADC hardware
+ * @brief Set an asynchronous read request.
  *
- * This routine disables the ADC hardware block for data sampling for the
- * specified device.
- *
- * @param dev Pointer to the device structure for the driver instance.
+ * @param config Pointer to a valid adc_config structure instance.
+ * @param sequence Pointer to an array with definitions of samplings to perform.
+ * @param sequence_length Number of entries in the sampling sequence array.
+ * @param async A pointer to a valid and ready to be signaled
+ *        struct k_poll_signal. (Note: if NULL this function will not
+ *        notify the end of the transaction, and whether it went
+ *        successfully or not).
  *
- * @return N/A
+ * @returns 0 on success, negative errno code otherwise.
  */
-__syscall void adc_disable(struct device *dev);
-
-static inline void _impl_adc_disable(struct device *dev)
+__syscall int adc_read_async(const struct adc_config *config,
+			     const struct adc_seq_entry *sequence,
+			     u8_t sequence_length,
+			     struct k_poll_signal *async);
+
+static inline int _impl_adc_read_async(const struct adc_config *config,
+				       const struct adc_seq_entry *sequence,
+				       u8_t sequence_length,
+				       struct k_poll_signal *async)
 {
-	const struct adc_driver_api *api = dev->driver_api;
+	const struct adc_driver_api *api = config->dev->driver_api;
 
-	api->disable(dev);
+	return api->read_async(config, sequence, sequence_length, async);
 }
+#endif /* CONFIG_ADC_ASYNC */
 
 /**
- * @brief Set a read request.
- *
- * This routine sends a read or sampling request to the ADC hardware block.
- * A sequence table describes the read request.
- * The routine returns once the ADC completes the read sequence.
- * The sample data can be retrieved from the memory buffers in
- * the sequence table structure.
+ * @brief Set an asynchronous read request.
```

Comment:
This comment seems incorrect.

Reply:
We should document the following - is it required to call adc_channel_setup before adc_read*?